### PR TITLE
[keyvault/internal] Add KeyVaultChallengePolicyOptions

### PR DESCRIPTION
**Pull request by chlowell**

### CODE REVIEW

**Review by jhendrixMSFT — APPROVED** [review]
I can't comment on the correctness of the algorithm, but the change itself looks good.  Just need to be careful on release timing as this is a breaking change and would break older KV clients.

File: sdk/keyvault/internal/challenge_policy.go [hunk 1]
```diff
@@ -27,18 +27,29 @@ import (
 const headerAuthorization = "Authorization"
 const bearerHeader = "Bearer "
 
+type KeyVaultChallengePolicyOptions struct {
+	// DisableChallengeResourceVerification controls whether the policy requires the
+	// authentication challenge resource to match the Key Vault or Managed HSM domain
+	DisableChallengeResourceVerification bool
```
Comment (RickWinter) [inline sdk/keyvault/internal/challenge_policy.go:33]: Is every language using this name?

File: sdk/keyvault/internal/challenge_policy.go [hunk 2]
```diff
@@ -27,18 +27,29 @@ import (
...
 	// mainResource is the resource to be retrieved using the tenant specified in the credential
-	mainResource *temporal.Resource[azcore.AccessToken, acquiringResourceState]
-	cred         azcore.TokenCredential
-	scope        *string
-	tenantID     *string
+	mainResource            *temporal.Resource[azcore.AccessToken, acquiringResourceState]
+	cred                    azcore.TokenCredential
+	scope                   *string
+	tenantID                *string
+	verifyChallengeResource bool
```
Comment (RickWinter) [inline sdk/keyvault/internal/challenge_policy.go:42]: I suggest dropping the *Resource* postfix.  We have that on mainResource which is a Resource.  In this case this is a switch and not a resource.

```suggestion
	verifyChallenge bool
```

File: sdk/keyvault/internal/challenge_policy.go (hunk 1 above)
Comment (heaths) [reply sdk/keyvault/internal/challenge_policy.go:33]: Except for Python, yes, we plan to.

File: sdk/keyvault/internal/challenge_policy.go (hunk 2 above)
Comment (heaths) [reply sdk/keyvault/internal/challenge_policy.go:42]: We are specifically verifying the `resource` in the challenge, not the entire challenge itself. Dropping "resource" is misleading.

**Review by heaths — APPROVED** [review]

**Review by heaths — APPROVED** [review]

**Review by heaths — APPROVED** [review]

**Pull request merged**